cache.go

```diff
@@ -57,6 +57,10 @@ func (c *cache) Set(k string, x interface{}, d time.Duration) {
 	if d > 0 {
 		e = time.Now().Add(d).UnixNano()
 	}
+
+	// delete before setting and call onEvicted
+	c.deleteIfExpired(k)
```

Comment:
@jaimem88 

Unfortunately, this has race condition, but additionally is pretty slow: multiple locks and multiple map searches.

We need to atomically swap item in a lock.

Something like this:

```golang
var item Item
var evicted bool

c.mu.Lock()
if c.onEvicted != nil {
  item, evicted := c.items[k]
}
c.items[k] = Item{
	Object:     x,
	Expiration: e,
}
// TODO: Calls to mu.Unlock are currently not deferred because defer
// adds ~200 ns (as of go1.)
c.mu.Unlock()

if evicted {
	c.onEvicted(k, item.Object)
}
```

Reply:
thanks @ayufan, I've updated it 😄

Reply:
@jaimem88

I think it is also possible that the same item is being set, because we update expiry time.

I think we need:

```golang
if evicted && item.Object != x {
	c.onEvicted(k, item.Object)
}
```

Reply:
@ayufan I think we want to call it if `item.Expired()`? which is the original issue #48  😅

Reply:
updated :)